### PR TITLE
ZIP 321 Reference Implementation

### PR DESCRIPTION
**Pull request by nuttycom**
Reference implementation for https://github.com/zcash/zips/pull/395

### CODE REVIEW

File: zcash_client_backend/src/zip321.rs [hunk 1]
```diff
@@ -0,0 +1,911 @@
...
+
+        // Parse the remaining parameters as an undifferentiated list
+        let (_, xs) = all_consuming(preceded(
+            char('?'),
+            separated_list(char('&'), parse::zcashparam(params)),
+        ))(rest)
+        .map_err(|e| e.to_string())?;
+
+        // Construct sets of payment parameters, keyed by the payment index.
+        let mut params_by_index: HashMap<usize, Vec<parse::Param>> = HashMap::new();
```
[the remark this answers is not in the document]
Comment (daira) [reply zcash_client_backend/src/zip321.rs:247]: [There can be up to 2109 outputs in a transaction](https://zips.z.cash/protocol/canopy.pdf#saplingbalance), so let's make it `"." NONZERO 0*3DIGIT`.

File: zcash_client_backend/src/zip321.rs [hunk 2]
```diff
@@ -0,0 +1,911 @@
...
+        Other(String, String),
+    }
+
+    #[derive(Debug)]
+    pub struct IndexedParam {
+        pub param: Param,
+        pub payment_index: usize,
+    }
+
+    pub fn has_param(v: &Vec<Param>, p: &Param) -> bool {
```
Comment (daira) [inline zcash_client_backend/src/zip321.rs:399]:
```suggestion
    pub fn has_duplicate_param(v: &Vec<Param>, p: &Param) -> bool {
```

File: zcash_client_backend/src/zip321.rs [hunk 3]
```diff
@@ -0,0 +1,911 @@
...
+                char('.'),
+                recognize(tuple((one_of("123456789"), digit0))),
+            )),
+        ))(input)
+    }
+
+    pub fn parse_amount<'a>(input: &'a str) -> IResult<&'a str, Amount> {
+        map_res(
+            tuple((digit1, opt(preceded(char('.'), digit1)))),
+            |(whole_s, decimal_s): (&str, Option<&str>)| {
```
Comment (daira) [inline zcash_client_backend/src/zip321.rs:517]: Please add a test case where the whole part is greater than the maximum `i64`.

File: zcash_client_backend/src/zip321.rs [hunk 4]
```diff
@@ -0,0 +1,911 @@
...
+                    .to_string()
+                    .parse::<i64>()
+                    .map_err(|e| e.to_string())?;
+
+                let zats: i64 = match decimal_s {
+                    Some(d) => d
+                        .char_indices()
+                        .map(|(i, c)| {
+                            let digit: i64 = c.to_digit(10).unwrap().into();
+                            digit * 10i64.pow((7 - i).try_into().unwrap())
```
Comment (daira) [inline zcash_client_backend/src/zip321.rs:528]: Try extending the string to 8 digits and then parsing.

File: zcash_client_backend/src/zip321.rs [hunk 5]
```diff
@@ -0,0 +1,911 @@
...
+    ) -> Result<IndexedParam, String> {
+        let param = match name {
+            "address" => Address::parse(params, value)
+                .map_err(|e| match e {
+                    AddrParseError::TAddrError(_) => format!(
+                        "Could not interpret {} as a valid transparent address.",
+                        value
+                    ),
+                    AddrParseError::SAddrError(_) => {
+                        format!("Could not interpret {} as a valid sapling address.", value)
```
Comment (daira) [inline zcash_client_backend/src/zip321.rs:555]:
```suggestion
                        format!("Could not interpret {} as a valid Sapling address.", value)
```

Comment (daira) [reply zcash_client_backend/src/zip321.rs:555]: Should we produce a more specific error for something that looks like a Sprout address (by prefix)?

File: zcash_client_backend/src/zip321.rs [hunk 6]
```diff
@@ -0,0 +1,911 @@
...
+        let mut query_params = vec![];
+        if self.payments.len() == 1 {
+            self.payments.get(0).map(|payment| {
+                push_params(payment, None, &mut query_params);
+                format!(
+                    "zcash:{}?{}",
+                    payment.recipient_address.render(params),
+                    query_params.join("&")
+                )
+            })
```
[the remark this answers is not in the document]
Comment (nuttycom) [reply zcash_client_backend/src/zip321.rs:218]: We already have:

> A ZIP-321 URI represents a request for the construction of a transaction having
> one or more *payments*. In the case that only a single payment is being
> requested, the recipient address MAY be included in the ``hier-part`` component
> of the RFC 3986 URI; otherwise, multiple recipient addresses can be specified
> using ``addrparam`` parameters with different indices.

Perhaps we should just change that MAY to SHOULD?

File: zcash_client_backend/src/zip321.rs (hunk 3 above)
Comment (nuttycom) [reply zcash_client_backend/src/zip321.rs:523]: https://github.com/zcash/librustzcash/pull/294/commits/32e85a65ea2733857a7d22df904d25a8798e4dab and the following two commits resolve this.

File: zcash_client_backend/src/zip321.rs [hunk 7]
```diff
@@ -0,0 +1,911 @@
...
+    amount: Amount,
+    memo: Option<Memo>,
+    label: Option<String>,
+    message: Option<String>,
+    other_params: Vec<(String, String)>,
+}
+
+impl Zip321Payment {
+    pub fn normalize(&mut self) {
+        self.other_params.sort();
```
[the remark this answers is not in the document]
Comment (nuttycom) [reply zcash_client_backend/src/zip321.rs:116]: This is only used in tests. I think probably the best thing to do here is make it pub(crate) and call it a day. Normalization isn't defined by the ZIP, there's no need for it to be part of the public API.

File: zcash_primitives/src/note_encryption.rs [hunk 8]
```diff
@@ -80,6 +81,27 @@ impl PartialEq for Memo {
...
+impl PartialOrd for Memo {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        let self_utf8 = self.to_utf8();
+        let other_utf8 = other.to_utf8();
+        match (self_utf8, other_utf8) {
+            (Some(Ok(a)), Some(Ok(b))) => a.partial_cmp(&b),
+            (Some(Ok(_)), _) => Some(Ordering::Less),
+            (_, Some(Ok(_))) => Some(Ordering::Greater),
+            (_, _) => self.0[..].partial_cmp(&other.0[..]),
+        }
```
Comment (daira) [inline zcash_primitives/src/note_encryption.rs:95]: Memos are allowed to be non-UTF-8 if they start with a byte >= 0xF5. In any case the validity is not enforced by consensus.

File: zcash_client_backend/src/zip321.rs [hunk 9]
```diff
@@ -0,0 +1,955 @@
...
+        move |input| {
+            input.split_at_position_complete(|item| {
+                let c = item.as_char();
+                !(c.is_alphanum() || allowed.contains(c))
+            })
+        }
+    }
+
+    pub fn qchars(input: &str) -> IResult<&str, &str> {
+        alphanum_or("-._~!$'()*+,;:@%".to_string())(input)
```
[the remark this answers is not in the document]
Comment (nuttycom) [reply zcash_client_backend/src/zip321.rs:503]: We already limit to essentially alphanumeric parameter names in the ZIP.

File: zcash_client_backend/src/zip321.rs [hunk 10]
```diff
@@ -0,0 +1,955 @@
...
+    }
+
+    pub fn indexed_name<'a>(input: &'a str) -> IResult<&'a str, (&'a str, Option<&'a str>)> {
+        let paramname = recognize(tuple((alpha1, namechars)));
+
+        tuple((
+            paramname,
+            opt(preceded(
+                char('.'),
+                recognize(tuple((one_of("123456789"), digit0))),
```
[the remark this answers is not in the document]
Comment (nuttycom) [reply zcash_client_backend/src/zip321.rs:517]: Good catch, I changed the zip but not the impl.

File: zcash_client_backend/src/zip321.rs [hunk 11]
```diff
@@ -0,0 +1,955 @@
...
+        .add(b']')
+        .add(b'^')
+        .add(b'`')
+        .add(b'{')
+        .add(b'|')
+        .add(b'}');
+
+    pub fn param_index(idx: Option<usize>) -> String {
+        match idx {
+            Some(i) if i > 0 => format!(".{}", i),
```
[the remark this answers is not in the document]
Comment (nuttycom) [reply zcash_client_backend/src/zip321.rs:325]: Since the type permits it, I prefer to keep the guard in place.

File: zcash_client_backend/src/zip321.rs [hunk 12]
```diff
@@ -0,0 +1,955 @@
...
+                    .map_err(|e| e.to_string())?;
+
+                let zats: i64 = match decimal_s {
+                    Some(d) => format!("{:0<8}", d)
+                        .parse::<i64>()
+                        .map_err(|e| e.to_string())?,
+                    None => 0,
+                };
+
+                let amt = coins * COIN + zats;
```
[the remark this answers is not in the document]
Comment (nuttycom) [reply zcash_client_backend/src/zip321.rs:538]: Amount::from_nonnegative_i64, just below here, takes care of  that.

File: zcash_primitives/src/note_encryption.rs (hunk 8 above)
Comment (nuttycom) [reply zcash_primitives/src/note_encryption.rs:95]: Yeah, this is also related to the source of the roundtrip errors; this is going to be reverted.

File: zcash_client_backend/src/zip321.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (daira) [reply zcash_client_backend/src/zip321.rs:503]: Oh, it's unambiguous already in the grammar that only ``qchar`` allows ``pct-encoded``. This is indeed allowed by my reading of RFC 3986. I'll add an explicit clarification and examples, though.

File: zcash_client_backend/src/zip321.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (daira) [reply zcash_client_backend/src/zip321.rs:538]: How so? I mean that the expression `coins * COIN + zats` can overflow the range of `i64`.

[the remark this answers is not in the document]
Comment (nuttycom) [reply zcash_client_backend/src/zip321.rs:609]: The limitation to MAX_MONEY, at least, is taken care of by the from_nonnegative_i64 call. Parse::<i64> fails in the case of values that are too large for an i64. But I do see the point, a weird URI could use the multiplication by COIN to overflow, and even though negative values are excluded by the later parse, I'd have to do the math to see if it could overflow all the way back in to small positive values; maybe it's better to just explicitly limit the coins value here. I'll limit the zats value to at most 8 characters in the parser, so that will eliminate that possibility here.

File: zcash_client_backend/src/zip321.rs [hunk 13]
```diff
@@ -578,21 +576,30 @@ pub mod parse {
...
         let paramname = recognize(tuple((alpha1, namechars)));
 
         tuple((
             paramname,
             opt(preceded(
                 char('.'),
-                recognize(tuple((one_of("123456789"), digit0))),
+                recognize(tuple((
+                    one_of("123456789"),
+                    map_opt(digit0, |s: &str| if s.len() > 3 { None } else { Some(s) }),
```
Comment (daira) [inline zcash_client_backend/src/zip321.rs:588]: Isn't [take_while_m_n!](https://docs.rs/nom/5.1.2/nom/macro.take_while_m_n.html) a clearer way to do this?

File: zcash_client_backend/src/zip321.rs [hunk 14]
```diff
@@ -606,6 +613,13 @@ pub mod parse {
                     None => 0,
                 };
 
+                if coins > 21000000 {
```
Comment (daira) [inline zcash_client_backend/src/zip321.rs:616]:
```suggestion
                if coins > 21000000 || (coins == 21000000 && zats > 0) {
```

File: zcash_client_backend/src/zip321.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (daira) [reply zcash_client_backend/src/zip321.rs:623]: Strictly speaking, `21000000.1` for example is not rejected when it should be.

File: zcash_client_backend/src/zip321.rs (hunk 13 above)
Comment (nuttycom) [reply zcash_client_backend/src/zip321.rs:588]: That's a byte-level combinator, not a character-level one. I actually tried using it here, but the cascade of replacing `&str` with `&[u8]` actually made for a lot more encoding and decoding, so I decided this was cleaner.

**Review by daira — APPROVED** [review]
utACK for the implementation. I have not looked at the tests in detail.

File: zcash_client_backend/src/zip321.rs [hunk 15]
```diff
@@ -0,0 +1,1046 @@
...
+            "amount" => parse_amount(value)
+                .map(|(_, a)| Param::Amount(a))
+                .map_err(|e| e.to_string()),
+
+            "label" => percent_decode(value.as_bytes())
+                .decode_utf8()
+                .map(|s| Param::Label(s.into_owned()))
+                .map_err(|e| e.to_string()),
+
+            "message" => percent_decode(value.as_bytes())
```
Comment (str4d) [inline zcash_client_backend/src/zip321.rs:662]: I dislike `message` as a parameter name, because it seems very likely to be confused with `memo`: users looking at a URI may think that `message` _is_ the memo. This could maybe be rectified by not encoding `message` in a human-readable format (e.g. also base64 it), but I think that defeats the intended purpose of `message`, so renaming the parameter makes more sense to me.

`label` is similarly non-descriptive and prone to confusion with `message`. I note that AFAICT the spec requires a `message` to be associated with a `recipient`, despite describing it as "descriptive information about the transaction". It feels like the overall intent is for `label` to be descriptions of recipients, and `message` is for a decription of the whole transaction, so maybe they should be named to indicate this?

File: zcash_client_backend/src/zip321.rs [hunk 16]
```diff
@@ -0,0 +1,1046 @@
...
+pub fn memo_to_base64(memo: &RawMemo) -> String {
+    memo.to_base64()
+}
+
+pub fn memo_from_base64(s: &str) -> Result<RawMemo, MemoError> {
+    RawMemo::from_base64(s)
+}
+
+#[derive(Debug, PartialEq)]
+pub struct Zip321Payment {
```
Comment (str4d) [inline zcash_client_backend/src/zip321.rs:172]: The naming of this struct confuses me. Isn't the intent of a ZIP 321 URI to request a single payment to potentially multiple recipients? So this would be something like `Recipient` or `RequestedRecipient`.

Separately, we should [avoid redundant prefixes](https://doc.rust-lang.org/1.0.0/style/style/naming/README.html#avoid-redundant-prefixes-[rfc-356]); these can be referred to as `zip321::PaymentRequest` etc. anywhere the prefix matters.

File: zcash_client_backend/src/zip321.rs [hunk 17]
```diff
@@ -0,0 +1,1046 @@
...
+                .then(a.memo.cmp(&b.memo))
+                .then(a.label.cmp(&b.label))
+                .then(a.message.cmp(&b.message))
+                .then(a.other_params.cmp(&b.other_params))
+        }
+    }
+}
+
+#[derive(Debug, PartialEq)]
+pub struct Zip321Request {
```
Comment (str4d) [inline zcash_client_backend/src/zip321.rs:211]:
```suggestion
pub struct PaymentRequest {
```

File: zcash_client_backend/src/zip321.rs [hunk 18]
```diff
@@ -0,0 +1,1046 @@
...
+    RawMemo::from_base64(s)
+}
+
+#[derive(Debug, PartialEq)]
+pub struct Zip321Payment {
+    recipient_address: Address,
+    amount: Amount,
+    memo: Option<RawMemo>,
+    label: Option<String>,
+    message: Option<String>,
```
Comment (str4d) [inline zcash_client_backend/src/zip321.rs:177]: Per my previous comment, I think `message` should be part of the `PaymentRequest` struct.

File: zcash_client_backend/src/zip321.rs (hunk 17 above)
Comment (str4d) [reply zcash_client_backend/src/zip321.rs:211]: Add documentation to (what are currently named) `Zip321Request` and `Zip321Payment`.

The other `pub struct`s and `pub enum`s will be refactored out of existence, so documenting them right now is non-blocking.

File: zcash_client_backend/src/zip321.rs [hunk 19]
```diff
@@ -0,0 +1,1046 @@
...
+        for (i, params) in params_by_index {
+            let payment = parse::to_payment(params, i)?;
+            payments.push(payment);
+        }
+
+        Ok(Zip321Request { payments })
+    }
+}
+
+pub mod render {
```
Comment (str4d) [inline zcash_client_backend/src/zip321.rs:356]: I don't see why this needs to be exposed beyond its use in `Zip321Request::to_uri`
```suggestion
mod render {
```

File: zcash_client_backend/src/zip321.rs [hunk 20]
```diff
@@ -0,0 +1,1046 @@
...
+        format!(
+            "{}{}={}",
+            label,
+            param_index(idx),
+            utf8_percent_encode(value, QCHAR_ENCODE)
+        )
+    }
+}
+
+pub mod parse {
```
Comment (str4d) [inline zcash_client_backend/src/zip321.rs:441]: Ditto:
```suggestion
mod parse {
```

File: zcash_client_backend/src/zip321.rs [hunk 21]
```diff
@@ -0,0 +1,1046 @@
...
+    use core::fmt::Debug;
+
+    use nom::{
```
Comment (str4d) [inline zcash_client_backend/src/zip321.rs:444]: For the `render` submodule, you might be interested in [`cookie-factory`](https://crates.io/crates/cookie-factory), which produces data in the same way that `nom` consumes it 😋🍪🏭

**Review by str4d — COMMENTED** [review]
Partial review; flushing comments.

File: zcash_client_backend/src/zip321.rs [hunk 22]
```diff
@@ -0,0 +1,1046 @@
...
+};
+use zcash_primitives::{consensus, legacy, primitives, transaction::components::Amount};
+
+use crate::encoding::{
+    decode_payment_address, decode_transparent_address, encode_payment_address,
+    encode_transparent_address,
+};
+
+#[derive(Debug, PartialEq, Clone)]
+pub enum Address {
```
[the remark this answers is not in the document]
Comment (nuttycom) [reply zcash_client_backend/src/zip321.rs:19]: That move is done as part of the data access API PR as well, so I'll just reproduce it here.

File: zcash_client_backend/src/zip321.rs [hunk 23]
```diff
@@ -0,0 +1,1046 @@
...
+    fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
+        f.debug_struct("RawMemo")
+            .field("memo", &format!("{:?}...", &self.0[0..17]))
+            .finish()
+    }
+}
+
+impl PartialEq for RawMemo {
+    fn eq(&self, other: &Self) -> bool {
+        self.to_base64() == other.to_base64()
```
[the remark this answers is not in the document]
Comment (nuttycom) [reply zcash_client_backend/src/zip321.rs:72]: Almost certainly not, I just didn't know about that. Still getting my head around Rust slice vs. array semantics.

File: zcash_client_backend/src/zip321.rs (hunk 16 above)
Comment (nuttycom) [reply zcash_client_backend/src/zip321.rs:115]: We define `Payment` in ZIP-321 as "A transfer of funds implemented by a shielded or transparent output of a single Zcash transaction." We needed a word for that, but perhaps there's a better one. "recipient" doesn't quite capture it, I don't think. Anyway this is consistent with the ZIP.

File: zcash_client_backend/src/zip321.rs (hunk 18 above)
Comment (nuttycom) [reply zcash_client_backend/src/zip321.rs:120]: So, if you want to display an itemized bill, you need a message per line item; however, I can see the argument for having a single "overall" message for the constructed transaction. Also this makes me think that I should add a "SHOULD" for how labels are treated by wallets (should be displayed preferentially instead of addresses if present.)

File: zcash_primitives/src/transaction/components/amount.rs [hunk 24]
```diff
@@ -147,6 +147,19 @@ impl Sum for Amount {
...
 }
 
+#[cfg(any(test, feature = "test-dependencies"))]
+pub mod testing {
+    use proptest::prelude::*;
+
+    use super::{Amount, MAX_MONEY};
+
+    prop_compose! {
+        pub fn arb_amount()(amt in 0i64..MAX_MONEY) -> Amount {
```
[the remark this answers is not in the document]
Comment (nuttycom) [reply zcash_primitives/src/transaction/components/amount.rs:157]: Using `arb_nonnegative_amount`.

File: zcash_client_backend/src/zip321.rs (hunk 21 above)
Comment (nuttycom) [reply zcash_client_backend/src/zip321.rs:386]: See, a proper schema library would only make you define the schema once... https://github.com/nuttycom/xenomorph

File: zcash_client_backend/src/zip321.rs (hunk 15 above)
Comment (nuttycom) [reply zcash_client_backend/src/zip321.rs:604]: This is actually taken from BIP-21. I will clarify the spec with respect to each of these though.

File: zcash_client_backend/proptest-regressions/zip321.txt [hunk 25]
```diff
@@ -0,0 +1,8 @@
+# Seeds for failure cases proptest has generated in the past. It is
+# automatically read and these particular cases re-run before any
+# novel cases are generated.
+#
+# It is recommended to check this file in to source control so that
+# everyone who runs the test benefits from these saved cases.
+cc 150e79b7e542f49689fad3c6e2bb3f830a78e4f2035e9fbab094d62e65f361d5 # shrinks to req = Zip321Request { payments: [Zip321Payment { recipient_address: SaplingAddress(PaymentAddress { pk_d: SubgroupPoint(ExtendedPoint { u: 0x25621e6b6515e85443822759660e9eef04918451ac4cd1263b8087d00689fdff, v: 0x11baaa27bc00dd67e8e03c98ca86c14b1f3ac420242f8e5cc5f9db1dced59b64, z: 0x168e654728395ce631656f756b4ae31dd19b6be3a31f55fd801f8272c5eb1dac, t1: 0x312ff1fd7ffb718f705f340bee2d80a6a0182efd81dd8f2a78cc468320be470f, t2: 0x0af52c1864feef43acb492b8bb106570ecca0ecd73567d8b2888769bb6022564 }), diversifier: Diversifier([59, 246, 250, 31, 131, 191, 69, 99, 200, 167, 19]) }), amount: Amount(0), memo: None, label: None, message: Some(""), other_params: [] }] }
```
Comment (zancas) [inline zcash_client_backend/proptest-regressions/zip321.txt:7]: Is there a best practice around formatting this kind of test vector?
This is my first look at `proptest` and it seems like making this human readable would be helpful.
I just checked:  https://github.com/zcash/librustzcash/pull/294/commits/ff1ecaf950cceea8d83a536226a8667af15c7e58

It seems like this comment is still relevant.

Comment (zancas) [reply zcash_client_backend/proptest-regressions/zip321.txt:7]: I think I will tinker a bit with formats...

File: zcash_client_backend/src/zip321.rs [hunk 26]
```diff
@@ -113,43 +112,67 @@ impl Memo {
...
+pub fn memo_from_base64(s: &str) -> Result<RawMemo, MemoError> {
+    RawMemo::from_base64(s)
+}
+
 #[derive(Debug, PartialEq)]
 pub struct Zip321Payment {
     recipient_address: Address,
     amount: Amount,
-    memo: Option<Memo>,
+    memo: Option<RawMemo>,
```
Comment (zancas) [inline zcash_client_backend/src/zip321.rs:118]: Is there a strong consensus that this field should have the identifier `memo`?

File: zcash_client_backend/src/zip321.rs [hunk 27]
```diff
@@ -0,0 +1,927 @@
...
+            Ok(None)
+        } else {
+            let t_res = decode_transparent_address(
+                &params.b58_pubkey_address_prefix(),
+                &params.b58_script_address_prefix(),
+                addr_str,
+            )
+            .map(|t| t.map(Address::TransparentAddress))
+            .map_err(AddrParseError::TAddrError);
+
```
Comment (zancas) [inline zcash_client_backend/src/zip321.rs]: Could you short circuit with `return` here such that `s_res` is not always computed?

Is there a security guarantee that is enforced by running this `fn` in constant time?

File: zcash_client_backend/src/zip321.rs (hunk 18 above)
Comment (nuttycom) [reply zcash_client_backend/src/zip321.rs:120]: I've added this for clarification: https://github.com/zcash/zips/pull/406

File: zcash_client_backend/src/zip321.rs (hunk 27 above)
Comment (nuttycom) [reply zcash_client_backend/src/zip321.rs]: This has actually *just* been subsumed by moving the existing parsing in the zcash_client_sqlite crate up the hierarchy, so this has now been deleted. :)

File: zcash_client_backend/proptest-regressions/zip321.txt (hunk 25 above)
Comment (nuttycom) [reply zcash_client_backend/proptest-regressions/zip321.txt:7]: These are automatically generated by proptest, so it probably shouldn't be messed with.

File: zcash_client_backend/src/zip321.rs (hunk 26 above)
Comment (nuttycom) [reply zcash_client_backend/src/zip321.rs:118]: I think so.. do you have another suggestion?

File: zcash_client_backend/proptest-regressions/zip321.txt (hunk 25 above)
Comment (nuttycom) [reply zcash_client_backend/proptest-regressions/zip321.txt:7]: I should elaborate: as proptest generates samples, when a sample that fails the assertions is generated it is persisted to this file and then automatically checked on subsequent proptest runs (using the value of `cc` as the seed for randomization). So this is never human-edited at all, just saved after proptest finds a failure.

File: zcash_client_backend/src/zip321.rs (hunk 15 above)
Comment (str4d) [reply zcash_client_backend/src/zip321.rs:604]: Hmm, okay. I still don't like it, but If we're explicitly wanting BIP-21 compatibility then I guess this is fine.

File: zcash_client_backend/src/zip321.rs [hunk 28]
```diff
@@ -0,0 +1,992 @@
...
+            other => percent_decode(value.as_bytes())
+                .decode_utf8()
+                .map(|s| Param::Other(other.to_string(), s.into_owned()))
+                .map_err(|e| e.to_string()),
+        }?;
+
+        let payment_index = match iopt {
+            Some(istr) => istr
+                .to_string()
```
Comment (str4d) [inline zcash_client_backend/src/zip321.rs:625]:
```suggestion
```

File: zcash_client_backend/src/zip321.rs [hunk 29]
```diff
@@ -0,0 +1,992 @@
...
+        let v4r = TransactionRequest::from_uri(&TEST_NETWORK, &valid_4).unwrap();
+        assert_eq!(
+            v4r.payments.get(0).map(|p| p.amount),
+            Some(Amount::from_u64(2100000000000000u64).unwrap())
+        );
+    }
+
+    #[test]
+    fn test_zip321_spec_invalid_examples() {
+        // invalid; missing `address=` or `address.0=`
```
Comment (str4d) [inline zcash_client_backend/src/zip321.rs:872]:
```suggestion
        // invalid; missing `address=`
```

**Review by str4d — CHANGES_REQUESTED** [review]

File: zcash_client_backend/src/zip321.rs [hunk 30]
```diff
@@ -0,0 +1,992 @@
...
+
+            Some(format!("zcash:?{}", query_params.join("&")))
+        }
+    }
+
+    /// Parse the provided URI to a payment request value.
+    pub fn from_uri<P: consensus::Parameters>(params: &P, uri: &str) -> Result<Self, String> {
+        // Parse the leading zcash:<address>
+        let (rest, primary_addr_param) =
+            parse::lead_addr(&params.clone())(uri).map_err(|e| e.to_string())?;
```
[the remark this answers is not in the document]
Comment (nuttycom) [reply zcash_client_backend/src/zip321.rs:224]: Oh... this was a hangover from a time when `lead_addr` returned a closure. I gave up on trying to write this as all combinators because the borrow checker gave me such fits.

File: zcash_client_backend/src/zip321.rs [hunk 31]
```diff
@@ -0,0 +1,992 @@
...
+        addr: &RecipientAddress,
+        idx: Option<usize>,
+    ) -> String {
+        format!("address{}={}", param_index(idx), addr.encode(params),)
+    }
+
+    pub fn amount_str(amount: Amount) -> Option<String> {
+        if amount.is_positive() {
+            let coins = i64::from(amount) / COIN;
+            let zats = i64::from(amount) % COIN;
```
[the remark this answers is not in the document]
Comment (nuttycom) [reply zcash_client_backend/src/zip321.rs:326]: The reason I used u64 is that, for some reason, the value of COIN is i64.

File: zcash_client_backend/src/zip321.rs [hunk 32]
```diff
@@ -0,0 +1,992 @@
...
+            "message" => percent_decode(value.as_bytes())
+                .decode_utf8()
+                .map(|s| Param::Message(s.into_owned()))
+                .map_err(|e| e.to_string()),
+
+            "memo" => memo_from_base64(value)
+                .map(Param::Memo)
+                .map_err(|e| format!("Decoded memo was invalid: {:?}", e)),
+
+            other if other.starts_with("req-") => {
```
[the remark this answers is not in the document]
Comment (nuttycom) [reply zcash_client_backend/src/zip321.rs:580]: I think that that would be a distinct parameter in that case; i.e. `req-amount` is a new required parameter distinct from `amount`. The intent is that we don't anticipate needing to add new required parameters, but we have a way to do so should something like that arise.

[the remark this answers is not in the document]
Comment (nuttycom) [reply zcash_client_backend/src/zip321.rs:580]: If there are independent extensions to the standard that use the `req-` form, then clients that conform to the zip-321 spec will correctly reject these parameters that they don't recognize (and consequently will not silently do the wrong thing.) This approach is also taken by bip-21: https://en.bitcoin.it/wiki/BIP_0021

File: zcash_client_backend/proptest-regressions/zip321.txt (hunk 25 above)
Comment (zancas) [reply zcash_client_backend/proptest-regressions/zip321.txt:7]: Awesome!
I'm really excited to learn about `proptest`, BTW.
A few thoughts:
(1)   It seems like if we had an automated formatter that was also checked in to git, so we could verifiably convert from "native" format to "human-readable" and back, then we could just "punch a button" to get the relevant format.  (It seems like it might not be worth the effort, out-of-the box, but I guess that's a function of how much effort it is to write (among other things!).)  On the other hand it might be a handy test-utility.
(2)  What if on every iteration of a proptest there was a _very_ low probability (tunable) that it would record its `seed`?    <--  This probability would be independent of whether the test succeeded-or-failed.   This would provide a uniform distribution of seeds over time.   Then what if there were a class of tests that were known to fail _sometimes_ where that "sometimes" was unquantified/estimated.    Then the distribution (frequency) of failures in the "sometimes" tests would provide quantitative data.  (Actually I'm not sure I have come up with a very good example.)
  For example, what if there was a test that fails whenever some "transaction/block"? isn't gossiped about enough in some period of time.   That test could fail for various reasons, e.g. someone tripped over the ethernet cable, or the block was malformed and other nodes won't acknowledge it, etc.  The point is that the amount of time until the test fails would show timing information like:
  90% of the time, the test passes, 10% of the time the test (by design! fails).
  So then one day someone asks:   How often is two minutes insufficient for a FOO to be gossiped about?   we'd have an answer!

OK, so maybe all of the above is already discussed elsewhere, or maybe I should move the thought elsewhere.

File: zcash_primitives/Cargo.toml [hunk 33]
```diff
@@ -36,14 +36,17 @@ ripemd160 = { version = "0.9", optional = true }
...
 
 [dev-dependencies]
 criterion = "0.3"
+proptest = "0.10.1"
 hex-literal = "0.2"
 rand_xorshift = "0.2"
 
 [features]
 transparent-inputs = ["ripemd160", "secp256k1"]
+test-dependencies = ["proptest"]
```
Comment (zancas) [inline zcash_primitives/Cargo.toml:49]: I totz lifted this `test-dependencies` pattern from here.  It's pretty sweet.

File: zcash_primitives/src/keys.rs [hunk 34]
```diff
@@ -180,6 +180,30 @@ impl FullViewingKey {
     }
 }
 
+#[cfg(any(test, feature = "test-dependencies"))]
```
Comment (zancas) [inline zcash_primitives/src/keys.rs:183]: And this.

File: zcash_primitives/src/keys.rs [hunk 35]
```diff
@@ -180,6 +180,30 @@ impl FullViewingKey {
...
+pub mod testing {
+    use proptest::collection::vec;
+    use proptest::prelude::*;
```
Comment (zancas) [inline zcash_primitives/src/keys.rs:186]: I find it harder to understand code that uses the `use  A::...*` pattern.
Is there any chance you'd be willing to provide the bound Items explicitly?

File: zcash_client_backend/src/zip321.rs [hunk 36]
```diff
@@ -0,0 +1,982 @@
...
+pub struct RawMemo([u8; 512]);
+
+#[derive(Debug)]
+pub enum MemoError {
+    InvalidBase64(base64::DecodeError),
+    LengthExceeded(usize),
+}
+
+impl RawMemo {
+    // Construct a memo from a vector of bytes.
```
Comment (zancas) [inline zcash_client_backend/src/zip321.rs:24]: Is this doc comment in the right spot?

File: zcash_primitives/src/keys.rs (hunk 35 above)
Comment (nuttycom) [reply zcash_primitives/src/keys.rs:186]: I have to admit that here I was just cargo-culting the proptest examples. I could try to figure out everything that's being used.

File: zcash_client_backend/src/zip321.rs (hunk 36 above)
Comment (nuttycom) [reply zcash_client_backend/src/zip321.rs:24]: Oops, good catch.

File: zcash_client_backend/src/zip321.rs [hunk 37]
```diff
@@ -0,0 +1,992 @@
...
+        .add(b'>')
+        .add(b'?')
+        .add(b'[')
+        .add(b'\\')
+        .add(b']')
+        .add(b'^')
+        .add(b'`')
+        .add(b'{')
+        .add(b'|')
+        .add(b'}');
```
[the remark this answers is not in the document]
Comment (nuttycom) [reply zcash_client_backend/src/zip321.rs:331]: @daira had also verified this earlier, if that helps.

**Review by str4d — APPROVED** [review]
